ffi/capi/src/calendar.rs

```diff
@@ -126,34 +128,33 @@ pub mod ffi {
         #[diplomat::rust_link(icu::calendar::AnyCalendar::try_new, FnInEnum)]
         #[diplomat::attr(supports = fallible_constructors, named_constructor = "for_locale_with_provider")]
         #[diplomat::demo(default_constructor)]
+        #[cfg(feature = "buffer_provider")]
         pub fn create_for_locale_with_provider(
             provider: &DataProvider,
             locale: &Locale,
         ) -> Result<Box<Calendar>, DataError> {
             let prefs = (&locale.0).into();
 
-            Ok(Box::new(Calendar(Arc::new(call_constructor!(
-                icu_calendar::AnyCalendar::try_new,
-                icu_calendar::AnyCalendar::try_new_with_any_provider,
-                icu_calendar::AnyCalendar::try_new_with_buffer_provider,
-                provider,
-                prefs
+            Ok(Box::new(Calendar(Arc::new(provider.call_constructor(
+                |provider| icu_calendar::AnyCalendar::try_new_with_buffer_provider(provider, prefs),
             )?))))
         }
 
         /// Creates a new [`Calendar`] from the specified date and time, using a particular data source.
         #[diplomat::rust_link(icu::calendar::AnyCalendar::new_for_kind, FnInEnum)]
         #[diplomat::attr(supports = fallible_constructors, named_constructor = "for_kind_with_provider")]
+        #[cfg(feature = "buffer_provider")]
         pub fn create_for_kind_with_provider(
             provider: &DataProvider,
             kind: AnyCalendarKind,
         ) -> Result<Box<Calendar>, DataError> {
-            Ok(Box::new(Calendar(Arc::new(call_constructor!(
-                icu_calendar::AnyCalendar::new_for_kind [r => Ok(r)],
-                icu_calendar::AnyCalendar::try_new_for_kind_with_any_provider,
-                icu_calendar::AnyCalendar::try_new_for_kind_with_buffer_provider,
-                provider,
-                kind.into()
+            Ok(Box::new(Calendar(Arc::new(provider.call_constructor(
+                |provider| {
+                    icu_calendar::AnyCalendar::try_new_for_kind_with_buffer_provider(
+                        provider,
+                        kind.into(),
+                    )
+                },
             )?))))
```

Comment:
I don't specifically recall a discussion on changing the macro to a closure. It seems like it would be harder in the future to change the behavior back to supporting `try_new_unstable` and similar. But it's fine.

Reply:
Nope, this PR was the discussion opportunity, and I can remove the last N commits if we don't want this.

But it should be easy enough to bring back when we need it.

I tend to consider macro use to be suboptimal, so if they can be removed without sacrificing anything I try to remove them. Here the complexity has not increased and the main downside is wanting to refactor again in the medium term.